File: LittleProjects/Scala/StreamFilterAkkaStreams/src/test/scala/stream/filter/akka/streams/InMemoryCache.scala
package stream.filter.akka.streams

import org.slf4j.LoggerFactory

import scala.collection.immutable
import scala.concurrent.{ExecutionContext, Future}

class InMemoryCache(p: Entity => Boolean) extends Cache {
  import InMemoryCache._

  override def isExist(entity: Entity): Boolean = {
    logger.debug(s"Check existence of $entity")
    p.apply(entity)
  }

  override def isExist(entities: immutable.Seq[Entity]): immutable.Seq[Boolean] = {
    logger.debug(s"Check existence of $entities")
    entities.map(p(_))
  }

  override def isExistAsync(entity: Entity)(implicit ex: ExecutionContext): Future[Boolean] = Future{
    logger.debug(s"Check existence of $entity")
    p.apply(entity)
  }

  override def isExistAsync(entities: immutable.Seq[Entity])(implicit ex: ExecutionContext): Future[immutable.Seq[Boolean]] = Future{
    logger.debug(s"Check existence of $entities")
    entities.map(p(_))
  }
}

object InMemoryCache {
  val logger = LoggerFactory.getLogger(InMemoryCache.getClass)

  def apply(p: Entity => Boolean): InMemoryCache = new InMemoryCache(p)
}

File: LittleProjects/Java/StreamFilteKafkaStreams/src/test/java/stream/filter/kafka/streams/InMemoryCache.java
package stream.filter.kafka.streams;

import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

public class InMemoryCache implements Cache {
    private Function<Entity, Boolean> predicate;

    public InMemoryCache(Function<Entity, Boolean> predicate) {
        this.predicate = predicate;
    }

    @Override
    public boolean isExist(Entity entity) {
        return predicate.apply(entity);
    }

    @Override
    public List<Boolean> isExistBatch(List<Entity> entities) {
        return entities.stream().map(e -> predicate.apply(e)).collect(Collectors.toList());
    }

    @Override
    public void close() {

    }
}
